SourceKit: correct exports attribution

compnerd opened the pull request:
Correct the DLL Storage for the library.  When building on ELF or MachO
targets, use visibility to ensure that the symbols are exported
properly.  This is required to build properly for Windows.

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

compnerd commented:
CC: @akyrtzi

compnerd commented:
@swift-ci please test

akyrtzi commented:
It's not clear to me why this change is necessary. sourcekitd implementation files are supposed to include `sourcekitd/Internal.h` at the beginning, which will define `SOURCEKITD_PUBLIC` as `__declspec(dllexport)`. Clients linking to sourcekitd will include `sourcekitd/sourcekitd.h` which defines `SOURCEKITD_PUBLIC` as `__declspec(dllimport)`.
Why is this scheme not working ?

compnerd commented:
@akyrtzi - you end up with differences in the declarations which causes a whole ton of spew.

akyrtzi commented:
> @akyrtzi - you end up with differences in the declarations which causes a whole ton of spew.

Isn't the following:
```
# if defined(sourcekitd_EXPORTS)
 #   define SOURCEKITD_PUBLIC __declspec(dllexport)
 # else
 #   define SOURCEKITD_PUBLIC __declspec(dllimport)
 # endif
```
doing the exact same thing ? `SOURCEKITD_PUBLIC` changes based on whether `sourcekitd_EXPORTS` is defined or not. The thing I'm not clear is what is the difference with the existing scheme ?
Is it maybe that there is an implementation file that is not including `sourcekitd/Internal.h` at the top ?

compnerd commented:
Yeah, it could be that there is a missing inclusion of `Internal.h`.  But, it can also be an issue if `Internal.h` and `sourcekitd.h` are included in the wrong order.  I'll see if I narrow down which files cause the warnings and if its a missing include.

compnerd committed:
SourceKit: correct header include order

Internal.h must be included before sourcekitd.h to ensure that we get the
correct definition for Windows targets.  The inclusion of `sourcekitd.h` is not
needed after the inclusion of `Internal.h` so simply remove the inclusion.

compnerd commented:
@swift-ci please test and merge

akyrtzi commented on tools/SourceKit/tools/sourcekitd/lib/API/sourcekitdAPI-InProc.cpp:
```diff
@@ -11,8 +11,8 @@
 #include "DictionaryKeys.h"
-#include "sourcekitd/sourcekitd.h"
 #include "sourcekitd/Internal.h"
+#include "sourcekitd/sourcekitd.h"
```
`sourcekitd/Internal.h` also includes `sourcekitd/sourcekitd.h` so you could remove the include of `sourcekitd/sourcekitd.h` from here completely.

compnerd commented:
@swift-ci please test and merge

compnerd commented:
@swift-ci please test and merge
